numpy/lib/tests/test_arraysetops.py

```diff
@@ -125,6 +128,60 @@ def test_ediff1d(self):
         assert_array_equal([7,1], ediff1d(two_elem, to_begin=7))
         assert_array_equal([5,6,1], ediff1d(two_elem, to_begin=[5,6]))
 
+    @pytest.mark.parametrize("ary, prepend, append", [
+        (np.array([1, 2, 3], dtype=np.int64),
+         None,
+         np.nan),
+        (np.array([1, 2, 3], dtype=np.int32),
+         np.array([5, 7, 2], dtype=np.int64),
+         None),
+        (np.array([1., 3., 9.], dtype=np.int8),
+         np.nan,
+         np.nan),
+         ])
+    def test_ediff1d_type_cast(self, ary, prepend, append):
+        # verify resolution of gh-11490
+
+        # specifically, raise an appropriate
+        # Exception when attempting to append or
+        # prepend with an incompatible type
+        msg = 'must be compatible'
+        with assert_raises_regex(TypeError, msg):
+            ediff1d(ary=ary,
+                    to_end=append,
+                    to_begin=prepend)
+
+    @pytest.mark.parametrize("ary,"
+                             "prepend,"
+                             "append,"
+                             "expected", [
+        (np.int16([1, 2, 3]),
+         0,
+         None,
+         np.array([0, 1, 1], dtype=np.int16)),
```

Comment:
Nit: A little inconsistent to spell the first of these `np.int16([1, 2, 3])` yet the latter `np.array([0, 1, 1], dtype=np.int16)` - would be better to pick one style and stick with it. The second is probably more typical